config_ext_test.go

```diff
@@ -0,0 +1,350 @@
+		var s string
+		val, err := provider.Get("practical.honda").WithDefault("CRV")
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: "civic",
+			Populate:  &s,
+		})
+		assert.Equal(t, "civic", s, "unexpected result after populate")
+	})
+
+	t.Run("default overriden by nil", func(t *testing.T) {
+		t.Skip("TODO: panics")
+		var s string
+		val, err := provider.Get("nothing").WithDefault("something")
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: nil,
+			Populate:  &s,
+		})
+		assert.Equal(t, "", s, "unexpected result after populate")
+	})
+
+	t.Run("default merges maps", func(t *testing.T) {
+		var m map[string]string
+		val, err := provider.Get("practical").WithDefault(map[string]string{"ford": "fiesta"})
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:    val,
+			HasValue: true,
+			Interface: map[interface{}]interface{}{
+				"honda":  "civic",
+				"toyota": "camry",
+				"nissan": "altima",
+				"ford":   "fiesta",
+			},
+			Populate: &m,
+		})
+		assert.Equal(t, map[string]string{
+			"honda":  "civic",
+			"toyota": "camry",
+			"nissan": "altima",
+			"ford":   "fiesta",
+		}, m, "unexpected result after populate")
+	})
+
+	t.Run("default replaces sequences", func(t *testing.T) {
+		var s []string
+		val, err := provider.Get("fun").WithDefault([]string{"delorean"})
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: []interface{}{"maserati", "lamborghini"},
+			Populate:  &s,
+		})
+		assert.Equal(t, []string{"maserati", "lamborghini"}, s, "unexpected result after populate")
+	})
+
+	t.Run("default ignores type mismatches", func(t *testing.T) {
+		// TODO: We should return an error in strict mode.
+		// fun is a sequence, so we shouldn't be able to merge in a mapping
+		val, err := provider.Get("fun").WithDefault(map[string]string{"dodge": "viper"})
+		assert.NoError(t, err, "error on type mismatch in permissive mode")
+		assert.Equal(t, []interface{}{"maserati", "lamborghini"}, val.Value(), "wrong merged output")
+	})
+
+	t.Run("chained defaults", func(t *testing.T) {
+		// Defaults should be FIFO: once we've set a default value, we can't later
+		// change our minds and set a different default.
+		val, err := provider.Get("top").WithDefault(map[string]string{"middle": "bottom"})
+		require.NoError(t, err, "couldn't set first default")
+		val, err = val.Get("middle").WithDefault("boom")
+		require.NoError(t, err, "couldn't set second default")
+
+		var s string
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: "bottom",
+			Populate:  &s,
+		})
+		assert.Equal(t, "bottom", s, "unexpected result after populate")
+	})
+
+	t.Run("deep copy", func(t *testing.T) {
+		t.Skip("TODO: mutations are visible to other callers")
```

Comment:
This is probably the most serious of the existing bugs.

Reply:
...oh no